test-framework/common/src/main/java/io/quarkus/test/common/TestResourceManager.java

```diff
@@ -459,4 +463,65 @@ public Annotation getConfigAnnotation() {
         }
     }
 
+    // visible for testing
+    static class DefaultTestInjector implements QuarkusTestResourceLifecycleManager.TestInjector {
+
+        // visible for testing
+        final Object testInstance;
+
+        private DefaultTestInjector(Object testInstance) {
+            this.testInstance = testInstance;
+        }
+
+        @Override
+        public void injectIntoFields(Object fieldValue, Class<?> fieldType, Class<? extends Annotation>... annotations) {
+            Class<?> c = testInstance.getClass();
+            while (c != Object.class) {
+                for (Field f : c.getDeclaredFields()) {
+                    AnnotationMatchResult annotationMatchResult = matchesAnnotations(f, annotations);
+                    if (annotationMatchResult != AnnotationMatchResult.NOT_MATCHED) {
+                        if (!fieldType.isAssignableFrom(f.getType())) {
+                            if (annotationMatchResult == AnnotationMatchResult.MATCHED) {
+                                // if the annotations match, we throw an exception if the types don't match as it's almost certainly a user error
+                                // if however there were no annotation supplied, then we continue searching for candidate fields
+                                throw new RuntimeException("'" + Arrays.toString(annotations)
```

Comment:
IIUC, this prevents some thing like:
```java
@InjectTestObject
Foo foo;

@InjectTestObject
Bar bar;
```
when `injectIntoFields(foo, Foo.class, InjectTestObject.class)` is called?

Isn't that too strict when a manager provides multiple objects that can be injected? I mean we would force users to create multiple annotations for that (or none at all).

Reply:
That is indeed true.

When writing this, I did get the feeling that the type parameter is likely unnecessary.
What do you think about making it an Optional?

Reply:
The thing is that then the method becomes difficult to use...

Reply:
I think I'm going to do something different. I'll have two methods: One which only takes an annotation as a param and one which takes a Predicate. That way if more advanced checking is needed, a user can specify it.

Reply:
Hmm, even though if it's more "hidden"/less expressive than using an annotation, injection simply by type is the easiest option after all because the user doesn't have to implement an annotation for that (KISS).

So I think (ideally) there should be four options:
- inject by type
- inject by annotation
- inject by type and annotation
- inject by whatever via Predicate

The first tree methods could simply call that Predicate method... 🤔

Reply:
And then you'll end up with a new CDI ;-). 

I think that `injectIntoFields(Object value, Predicate<Field> predicate)` is flexible enough.

Reply:
And I thought there is much more to CDI than that. 😉 
Anyway, my primary focus was to support lazy devs (aka developer joy) but yeah, you can do everything the Predicate.

Reply:
I updated the PR. WDYT?